File: new-version.py
import subprocess
import questionary
import os
import shutil
import sys
import toml
import re

# Background update check (optional)
try:
    script_dir = os.path.dirname(os.path.abspath(__file__))
    import update_check
    update_check.background_check(script_dir)
    flag_file = os.path.join(script_dir, ".update_available.json")
    if os.path.exists(flag_file):
        print("⚠️ Update available. Open the main utility and choose 'Check for updates' to update.")
except Exception:
    pass


def get_npm_executable():
    npm_exe = shutil.which("npm")
    if not npm_exe:
        print("❌ npm not found in PATH. Ensure Node.js is installed and available.")
        sys.exit(1)
    return npm_exe


def check_git_available():
    """Ensure git is available."""
    git_exe = shutil.which("git")
    if not git_exe:
        print("❌ git not found in PATH. Ensure Git is installed and available.")
        sys.exit(1)
    return git_exe


def check_git_clean():
    """Ensure the git working tree is clean."""
    try:
        result = subprocess.run(["git", "status", "--porcelain"], capture_output=True, text=True, check=True)
        if result.stdout.strip():
            print("❌ Git working tree contains uncommitted changes.")
            print("Please commit or discard changes before bumping the version.")
            sys.exit(1)
    except subprocess.CalledProcessError as e:
        print(f"❌ Error checking git status: {e}")
        sys.exit(1)


def git_commit_and_tag(version: str, message: str = None):
    """Create a git commit and tag for the new version."""
    try:
        subprocess.run(["git", "add", "pyproject.toml"], check=True)
        commit_message = message or f"v{version}"
        subprocess.run(["git", "commit", "-m", commit_message], check=True)
        print(f"✅ Created commit: {commit_message}")
        tag_name = f"v{version}"
        subprocess.run(["git", "tag", tag_name], check=True)
        print(f"✅ Created tag: {tag_name}")
    except subprocess.CalledProcessError as e:
        print(f"❌ Git operation failed: {e}")
        sys.exit(1)


def run_npm_version(script_name: str):
    npm_exe = get_npm_executable()
    try:
        subprocess.run([npm_exe, "version", script_name], check=True)
    except subprocess.CalledProcessError as e:
        print(f"❌ npm version {script_name} failed with exit code {e.returncode}")
        sys.exit(e.returncode)


def get_python_version(pyproject_path: str) -> str:
    """Read the project version from pyproject.toml."""
    with open(pyproject_path, "r", encoding="utf-8") as f:
        data = toml.load(f)

    version = (
        data.get("project", {}).get("version") or data.get("tool", {}).get("poetry", {}).get("version")
    )

    if not version:
        print("❌ Could not find version in pyproject.toml")
        sys.exit(1)

    return version


def bump_python_version(version: str, bump_type: str) -> str:
    """Return a bumped semantic version string."""
    match = re.match(r"(\d+)\.(\d+)\.(\d+)", version)
    if not match:
        print(f"❌ Invalid version format: {version}")
        sys.exit(1)

    major, minor, patch = map(int, match.groups())

    if bump_type == "patch":
        patch += 1
    elif bump_type == "minor":
        minor += 1
        patch = 0
    elif bump_type == "major":
        major += 1
        minor = 0
        patch = 0

    return f"{major}.{minor}.{patch}"


def update_pyproject_version(pyproject_path: str, new_version: str):
    """Update the version value inside pyproject.toml.

    This performs a conservative regex replace in either [project] or [tool.poetry]
    sections.
    """
    with open(pyproject_path, "r", encoding="utf-8") as f:
        content = f.read()

    updated = re.sub(
        r'(^\[project\].*?^version\s*=\s*")[^"]+(\")',
        rf'\g<1>{new_version}\g<2>',
        content,
        flags=re.MULTILINE | re.DOTALL,
    )

    if updated == content:
        updated = re.sub(
            r'(^\[tool\.poetry\].*?^version\s*=\s*")[^"]+(\")',
            rf'\g<1>{new_version}\g<2>',
            content,
            flags=re.MULTILINE | re.DOTALL,
        )

    if updated == content:
        print("❌ Failed to update version in pyproject.toml")
        sys.exit(1)

    with open(pyproject_path, "w", encoding="utf-8") as f:
        f.write(updated)


def run_python_version_bump(bump_type: str):
    pyproject_path = "pyproject.toml"
    check_git_available()
    check_git_clean()

    current_version = get_python_version(pyproject_path)
    print(f"📦 Current version: {current_version}")

    new_version = bump_python_version(current_version, bump_type)
    print(f"📦 New version: {new_version}")

    update_pyproject_version(pyproject_path, new_version)
    print("✅ Updated version in pyproject.toml")

    git_commit_and_tag(new_version)


def main():
    has_package_json = os.path.isfile("package.json")
    has_pyproject_toml = os.path.isfile("pyproject.toml")

    if not has_package_json and not has_pyproject_toml:
        print("❌ No package.json or pyproject.toml found in the current directory")
        sys.exit(1)

    project_type = None
    if has_package_json and has_pyproject_toml:
        project_type = questionary.select(
            "Both project types detected. Select one:",
            choices=["Node.js (package.json)", "Python (pyproject.toml)", "Exit"],
        ).ask()

        if not project_type or project_type.startswith("Exit"):
            print("Exiting...")
            return

        project_type = "npm" if "Node.js" in project_type else "python"
    elif has_package_json:
        project_type = "npm"
    else:
        project_type = "python"

    if project_type == "npm":
        get_npm_executable()

    choices = ["Patch (0.0.X)", "Minor (0.X.0)", "Major (X.0.0)", "Exit"]
    choice = questionary.select("Select version bump type:", choices=choices).ask()

    if not choice or choice.startswith("Exit"):
        print("Exiting...")
        return

    bump_type = choice.split(" ")[0].lower()

    if project_type == "npm":
        run_npm_version(bump_type)
    else:
        run_python_version_bump(bump_type)


if __name__ == "__main__":
    main()

File: src/core/new-version.ts
import fs from "node:fs";
import { spawnSync } from "node:child_process";
import inquirer from "inquirer";
import { backgroundCheck, notifyIfUpdateAvailable } from "./update-check";
import { header, info, warn, success, error, panel, kv, section, bullets, step } from "./utils/tui";
import { DetectApp } from "./utils/path";

const scriptDir = __dirname;

function commandExists(command: string) {
  const check = spawnSync(command, ["--version"], {
    stdio: "ignore",
    shell: process.platform === "win32",
  });
  return check.status === 0;
}

function runCommand(command: string, args: string[], options: { capture?: boolean } = {}) {
  const result = spawnSync(command, args, {
    encoding: "utf8",
    stdio: options.capture ? ["ignore", "pipe", "pipe"] : "inherit",
    shell: process.platform === "win32",
  });

  if ((result.error as any)?.code === "ENOENT") {
    throw new Error(`${command} not found in PATH.`);
  }

  if (result.status !== 0) {
    const stderr = result.stderr || `${command} ${args.join(" ")} failed.`;
    throw new Error(stderr.trim());
  }

  return result.stdout || "";
}

function ensureGitReady() {
  if (!commandExists("git")) {
    throw new Error("git not found in PATH.");
  }

  const status = runCommand("git", ["status", "--porcelain"], { capture: true }).trim();
  if (status.length > 0) {
    throw new Error("Git working tree has uncommitted changes.");
  }
}

function bumpSemver(version: string, bumpType: "patch" | "minor" | "major") {
  const match = version.match(/^(\d+)\.(\d+)\.(\d+)$/);
  if (!match) {
    throw new Error(`Invalid semantic version: ${version}`);
  }

  const major = Number.parseInt(match[1], 10);
  const minor = Number.parseInt(match[2], 10);
  const patch = Number.parseInt(match[3], 10);

  if (bumpType === "patch") {
    return `${major}.${minor}.${patch + 1}`;
  }

  if (bumpType === "minor") {
    return `${major}.${minor + 1}.0`;
  }

  if (bumpType === "major") {
    return `${major + 1}.0.0`;
  }

  throw new Error(`Unsupported bump type: ${bumpType}`);
}

function gitCommitAndTag(version: string, filesToAdd: string[]) {
  runCommand("git", ["add", ...filesToAdd]);
  runCommand("git", ["commit", "-m", `v${version}`]);
  runCommand("git", ["tag", `v${version}`]);
  success(`Created commit and tag for v${version}.`);
}

function updatePyprojectVersion(newVersion: string) {
  const pyprojectPath = "pyproject.toml";
  const content = fs.readFileSync(pyprojectPath, "utf8");

  let updated = content.replace(
    /(^\[project\][\s\S]*?^version\s*=\s*")[^"]+(")/m,
    `$1${newVersion}$2`
  );

  if (updated === content) {
    updated = content.replace(
      /(^\[tool\.poetry\][\s\S]*?^version\s*=\s*")[^"]+(")/m,
      `$1${newVersion}$2`
    );
  }

  if (updated === content) {
    throw new Error("Failed to update version in pyproject.toml.");
  }

  fs.writeFileSync(pyprojectPath, updated, "utf8");
}

function readPyprojectVersion() {
  const content = fs.readFileSync("pyproject.toml", "utf8");
  const projectMatch = content.match(/\[project\][\s\S]*?^version\s*=\s*"([^"]+)"/m);
  if (projectMatch?.[1]) {
    return projectMatch[1];
  }

  const poetryMatch = content.match(/\[tool\.poetry\][\s\S]*?^version\s*=\s*"([^"]+)"/m);
  if (poetryMatch?.[1]) {
    return poetryMatch[1];
  }

  throw new Error("Could not find version in pyproject.toml.");
}

function runPythonProjectBump(bumpType: "patch" | "minor" | "major") {
  ensureGitReady();
  const current = readPyprojectVersion();
  const next = bumpSemver(current, bumpType);
  info(`Current version: ${current}`);
  info(`New version: ${next}`);

  updatePyprojectVersion(next);
  gitCommitAndTag(next, ["pyproject.toml"]);
}

function runVersionFileBump(bumpType: "patch" | "minor" | "major") {
  ensureGitReady();
  const current = fs.readFileSync("VERSION", "utf8").trim();
  const next = bumpSemver(current, bumpType);
  info(`Current version: ${current}`);
  info(`New version: ${next}`);

  fs.writeFileSync("VERSION", `${next}\n`, "utf8");
  gitCommitAndTag(next, ["VERSION"]);
}

function runNpmVersionBump(bumpType: "patch" | "minor" | "major") {
  if (!commandExists("npm")) {
    throw new Error("npm not found in PATH.");
  }

  runCommand("npm", ["version", bumpType]);
}

async function runNewVersionMenu() {
  await backgroundCheck(scriptDir);
  notifyIfUpdateAvailable(scriptDir);
  header("TerminalUtils", "Version bump utility");
  panel("Release flow", [
    "Choose the project manifest to update.",
    "A clean git tree is required for pyproject.toml and VERSION flows.",
    "Version bumps create commit and tag where applicable.",
  ]);

  const hasPackageJson = fs.existsSync("package.json");
  const hasPyproject = fs.existsSync("pyproject.toml");
  const hasVersionFile = fs.existsSync("VERSION");

  if (!hasPackageJson && !hasPyproject && !hasVersionFile) {
    throw new Error("No package.json, pyproject.toml, or VERSION file found.");
  }

  section("Detected manifests", "Only available release targets are shown");
  bullets([
    `package.json: ${hasPackageJson ? "available" : "missing"}`,
    `pyproject.toml: ${hasPyproject ? "available" : "missing"}`,
    `VERSION: ${hasVersionFile ? "available" : "missing"}`,
  ]);

  const projectChoices = [];
  if (hasPackageJson) {
    projectChoices.push({ name: "Node.js (package.json)", value: "npm" });
  }
  if (hasPyproject) {
    projectChoices.push({ name: "Python (pyproject.toml)", value: "python" });
  }
  if (hasVersionFile) {
    projectChoices.push({ name: "Plain VERSION file", value: "version_file" });
  }
  projectChoices.push({ name: "Exit", value: "exit" });

  let projectType = projectChoices[0].value;
  if (projectChoices.length > 2) {
    const projectAnswer = await inquirer.prompt([
      {
        type: "list",
        name: "projectType",
        message: "Select project type:",
        choices: projectChoices,
      },
    ]);
    projectType = projectAnswer.projectType;
  }

  if (projectType === "exit") {
    info("Exiting.");
    return;
  }

  const bumpAnswer = await inquirer.prompt([
    {
      type: "list",
      name: "bumpType",
      message: "Select version bump type:",
      choices: [
        { name: "Patch (0.0.X)", value: "patch" },
        { name: "Minor (0.X.0)", value: "minor" },
        { name: "Major (X.0.0)", value: "major" },
        { name: "Exit", value: "exit" },
      ],
    },
  ]);

  if (bumpAnswer.bumpType === "exit") {
    info("Exiting.");
    return;
  }

  panel("Selected change", [
    kv("Project", projectType),
    kv("Bump", bumpAnswer.bumpType),
  ]);

  if (projectType === "npm") {
    step("Run npm version", bumpAnswer.bumpType);
    runNpmVersionBump(bumpAnswer.bumpType);
    success("npm version completed.");
    return;
  }

  if (projectType === "python") {
    step("Update pyproject.toml", bumpAnswer.bumpType);
    runPythonProjectBump(bumpAnswer.bumpType);
    return;
  }

  step("Update VERSION file", bumpAnswer.bumpType);
  runVersionFileBump(bumpAnswer.bumpType);
}

if (DetectApp() === "version") {
  runNewVersionMenu().catch((runError) => {
    error(runError.message);
    process.exit(1);
  });
}

export {
  runNewVersionMenu,
};
